### ISSUE TITLE
Unclear Degenerate coordinates error

### ISSUE THREAD

**Issue by nikosavola** [awslabs/palace#89]
*Description: A clear and concise description of what the bug is.*

A lumped-element driven full-wave simulation of mine  works for a previous version of Palace
>   "GitTag": "v0.11.2-14-g89810f9",

but on current main I get the following error, which feels a bit unclear

```
Verification failed: (&t_0 != &t_1) is false:
 --> Degenerate coordinates!
 ... in function: palace::mesh::BoundingBox palace::mesh::{anonymous}::BoundingBoxFromPointCloud(MPI_Comm, const std::vector<Eigen::Matrix<double, 3, 1> >&, int)
 ... in file: /opt/palace-src/palace/utils/geodata.cpp:787
```

I guess this is related to changes in https://github.com/awslabs/palace/pull/75, what is it exactly complaining about?

*To reproduce: Please provide minimal example that reproduces the error. For existing
examples, please provide a link.*

No MWE at least yet

*Error message: Paste the complete error message or log, if applicable.*
```
 _____________     _______
_____   __   \____ __   /____ ____________
____   /_/  /  __ ` /  /  __ ` /  ___/  _ \
___   _____/  /_/  /  /  /_/  /  /__/  ___/
  /__/     \___,__/__/\___,__/\_____\_____/

Running with 1 MPI process and 1 OpenMP thread

Finished partitioning mesh into 1 subdomain

Warning!
One or more external boundary attributes has no associated boundary condition!
"PMC"/"ZeroCharge" condition is assumed!

Removed 14088 unmarked domain elements from the mesh
Removed 1536 unattached boundary elements from the mesh

Characteristic length and time scales:
 L₀ = 2.700e-03 m, t₀ = 9.006e-03 ns

Mesh curvature order: 1
Mesh bounding box:
 (Xmin, Ymin, Zmin) = (-2.700e-03, -1.700e-03, +0.000e+00) m
 (Xmax, Ymax, Zmax) = (-3.750e-04, +1.000e-03, +1.760e-03) m

Coarse Parallel Mesh Stats:

                minimum     average     maximum       total
 vertices          5934        5934        5934        5934
 edges            32984       32984       32984       32984
 faces            48854       48854       48854       48854
 elements         21810       21810       21810       21810
 neighbors            0           0           0

            minimum     maximum
 h      0.000112883    0.119102
 kappa      1.05971      652186

Refined Parallel Mesh Stats:

                minimum     average     maximum       total
 vertices         38918       38918       38918       38918
 edges           234340      234340      234340      234340
 faces           369896      369896      369896      369896
 elements        174480      174480      174480      174480
 neighbors            0           0           0

            minimum     maximum
 h      5.64416e-05    0.059551
 kappa      1.05971      652192


Warning!
Unknown PEC boundary attribute 47!
Solver will just ignore it!


Configuring Robin absorbing BC (order 1) at attributes:
 53-54


Verification failed: (&t_0 != &t_1) is false:
 --> Degenerate coordinates!
 ... in function: palace::mesh::BoundingBox palace::mesh::{anonymous}::BoundingBoxFromPointCloud(MPI_Comm, const std::vector<Eigen::Matrix<double, 3, 1> >&, int)
 ... in file: /opt/palace-src/palace/utils/geodata.cpp:787

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
```
*Environment: Any environment details, such as operating system, compiler, etc.*
Singularity image built on current `main`

**Comment by hughcars** [comment]
Hi @nikosavola,

That line is triggering because the discovered set of points defining the bounding box are not returning as planar, but then the two discovered points (in and out of the plane) are referencing the same point. This could be a tolerance issue, or the lumped port itself is not defining a quadrilateral, or something else. I am guessing given it previously worked, it is not the second.

To check the tolerancing issue, could you please try changing the lines defining `box.planar` to:
```
    constexpr double planar_tolerance = 1.0e-6;
    box.planar = (std::abs(n_3.dot(t_0)) / n_3.norm() / t_0.norm()) < planar_tolerance &&
                 (std::abs(n_3.dot(t_1)) / n_3.norm() / t_1.norm()) < planar_tolerance;
```
and rebuilding?

**Comment by nikosavola** [comment]
I tried with your code snippet but the error persists…

I have two lumped-element ports that look like this on a CPW
![Screenshot 2023-08-23 15 43 40](https://github.com/awslabs/palace/assets/7860886/817f6e71-24f4-4865-beca-3737040ec91a)

![Screenshot 2023-08-23 15 43 21](https://github.com/awslabs/palace/assets/7860886/8b5999d9-a6f7-48f1-b409-109d90b187b8)

**Comment by hughcars** [comment]
Hi @nikosavola, could you try `hughcars/lumped-port-fix` and see if that works?

**Comment by hughcars** [comment]
Apologies, please try again now.

**Comment by nikosavola** [comment]
Now I get the following
```
Verification failed: (bounding_box.planar) is false:
 --> Boundary elements must be coplanar to define a lumped element!
 ... in function: palace::UniformElementData::UniformElementData(const std::array<double, 3>&, const mfem::Array<int>&, mfem::ParFiniteElementSpace&)
 ... in file: /opt/palace-src/palace/fem/lumpedelement.hpp:71
```
I'll try to provide an MWE mesh and settings

**Comment by hughcars** [comment]
That error message "Boundary elements must be coplanar to define a lumped element!" is indicating that the vertices that make up the lumped port have some non-negligible thickness, which is kind of surprising. It might be worth printing out the normals of the discovered box, it’s possible that the third dimension is very small and this is a case of numerical tolerancing. I had chosen the tolerance value based on some local experiments, but clearly it might not generalize very well. 

In lieu of a MWE, in the bounding box code just before this error is thrown, you could print out the contents of `vertices`, 
```
for (const auto &x : vertices)
{
  std::cout << x.transpose() << '\n';
}
```
then if you send that to me, I can work on the discovered set of points without needing access to a mesh.

**Comment by nikosavola** [comment]
I made a MWE where this happens on just a CPW. I put some thickness of 200 nm for the metal layer and tried to create the ports using the interface to vacuum surface. There are unused layers on the interfaces but they shouldn't matter.

[lumped_port_mwe.zip](https://github.com/awslabs/palace/files/12433986/lumped_port_mwe.zip)

**Comment by hughcars** [comment]
Visualizing the vertices from the mwe you had, I see the issue. There was a confusion on my part: I had been presuming that a lumped port would be defined as embedded within a plane, hence the check for planar status of the bounding box. This is not the whole story however, as there could be a topologically two dimensional surface embedded in three dimensional space (your port being a good example). This should be a valid port, and I am working on a fix.

Given you are working with a perfectly axis aligned port, I would suggest using `git checkout b9a080c2` from your local build. This should unblock you in the meantime.

Thank your for your understanding.

**Comment by nikosavola** [comment]
Hi @hughcars, appears to be working now, thanks!

**Comment by hughcars** [comment]
https://github.com/awslabs/palace/pull/94

**Comment by sebastiangrimberg** [comment]
Closed by #94
